src/common/offline_transformations/src/pruning/propagate_masks.cpp

```diff
@@ -582,20 +591,93 @@ class ngraph::pass::mask_propagation::PassThrough : public MatcherPass {
+                            cur_mask->at(dim) = input_mask_row->at(dim);
+                    return true;
+                }, input_mask);
+
+                // Invalidate current mask and its parent masks
+                output_mask->apply_callback(input_mask);
+                setMask(m_output, input_mask);
+            }
+
+            return true;
+        };
+
+        auto m = std::make_shared<ngraph::pattern::Matcher>(pooling_by_reduce, "PassThroughReduceMaskPropagation");
+        register_matcher(m, callback);
+    }
+};
+
 class ngraph::pass::mask_propagation::StopPropagation : public MatcherPass {
 public:
     StopPropagation() {
         auto any_node = pattern::any_input();
 
-        ngraph::matcher_pass_callback callback = [](ngraph::pattern::Matcher& m) {
+        ngraph::matcher_pass_callback callback = [=](ngraph::pattern::Matcher& m) {
+            const auto & pattern_map = m.get_pattern_value_map();
+            const auto & m_output = pattern_map.at(any_node);
             const auto & node = m.get_match_root();
+
+            auto output_mask = std::make_shared<Mask>(m_output.get_partial_shape().rank().get_length());
+            bool any_input_with_masks = false;
             for (const auto & input : node->input_values()) {
-                if (auto mask = getMask(input)) {
-                    // Invalidate current mask and its parent masks
-                    mask->invalidate();
-                    NGRAPH_DEBUG << "Invalidate masks for " << *input.get_node() << " because " << node << " is unknown\n";
+                if (auto input_mask = getMask(input)) {
+                        auto input_mask_row = input_mask.get();
+                        input_mask->add_callback([](Mask::Ptr cur_mask) -> bool {
+                            cur_mask->clean_dim_values();
+                            return true;
+                        }, output_mask);
+                        output_mask->add_callback([input_mask_row](Mask::Ptr cur_mask) -> bool{
+                            cur_mask->copy_value_from_mask(input_mask_row);
+                            return true;
+                        }, input_mask);
+
+                        // Invalidate current mask and its parent masks
+                        output_mask->apply_callback(input_mask);
+                        NGRAPH_DEBUG << "Invalidate masks for " << *input.get_node() << " because " << node << " is in scope of stop ops.\n";
+                        any_input_with_masks = true;
+                    }
                 }
+            if (any_input_with_masks) {
+                // TODO: find out if any other operation except Result
+                // share output tesnors with previous operation
+                if (ngraph::is_type<opset6::Result>(m_output.get_node_shared_ptr()))
+                    setResultMask(m_output, output_mask);
+                else
+                    setMask(m_output, output_mask);
```

Comment:
Have found a problem with `Result` operation: it share `m_tensor` attribute with previous operation, mask assign to `result` operation leads to previous operation mask override. Solve it by saving result mask by separate result operation key. Nobody needs to get mask from a result layer, but result mask should be saved to keep pruning masks in valid condition.

Reply:
CC: @GlebKazantaev